File: erlang/leap/src/leap.erl
-module(leap).

-export([leap_year/1]).

-define(LEAP_YEAR, 4).
-define(CENTURIAL_YEAR, 100).
-define(LEAP_CYCLE_LENGTH, (?LEAP_YEAR * ?CENTURIAL_YEAR)).


leap_year(Year) ->
  is_leap(Year) andalso (is_non_centurial(Year) orelse is_leap_cycle(Year)).

% Private

is_leap(Year) ->
  Year rem ?LEAP_YEAR == 0.

is_non_centurial(Year) ->
  Year rem ?CENTURIAL_YEAR /= 0.

is_leap_cycle(Year) ->
  Year rem ?LEAP_CYCLE_LENGTH == 0.

File: julia/leap/leap.jl
const LEAP_YEAR = 4
const CENTURIAL_YEAR = 100
const LEAP_CYCLE_LENGTH = LEAP_YEAR * CENTURIAL_YEAR

function is_leap_year(year::Int)
    is_leap(year) && (is_non_centurial(year) || is_leap_cycle(year))
end

function is_leap(year::Int)
    year % LEAP_YEAR == 0
end

function is_non_centurial(year::Int)
    year % CENTURIAL_YEAR != 0
end

function is_leap_cycle(year::Int)
    year % LEAP_CYCLE_LENGTH == 0
end
